### ISSUE TITLE
 tidb_ghpr_integration_copr_test fails with CONCAT in where clause

### ISSUE THREAD

**Issue by SunRunAway** [pingcap/tidb#16629]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```

[2020-04-20T03:52:58.721Z] Test case: sql/randgen-limit/7_str_1.sql

[2020-04-20T03:52:58.721Z] Statement: #514 -  SELECT UNHEX( ( CONCAT( `col_float` ) ) ) AS field1 FROM `table1_int_autoinc` WHERE CONCAT( ( '18:02:07.032977' / '1998-04-26' ), ( ~ `col_int` ) ) LIMIT 11 /* QNO 516 CON_ID 212 */ ;

[2020-04-20T03:52:58.721Z] NoPushDown Output: 

[2020-04-20T03:52:58.721Z] field1

[2020-04-20T03:52:58.721Z] $E

[2020-04-20T03:52:58.721Z] 

[2020-04-20T03:52:58.722Z] 

[2020-04-20T03:52:58.722Z] WithPushDown Output: 

[2020-04-20T03:52:58.722Z] field1
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_integration_copr_test/detail/tidb_ghpr_integration_copr_test/3843/pipeline

**Comment by SunRunAway** [comment]
NoPushDown:
```
mysql> explain analyze SELECT 1 FROM `table1_int_autoinc` WHERE CONCAT( ( '18:02:07.032977' / '1998-04-26' ), ( ~ `col_int` ) ) ;
+---------------------------+---------+---------+-----------+--------------------------+------------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
| id                        | estRows | actRows | task      | access object            | execution info                                                         | operator info                                                                 | memory    | disk |
+---------------------------+---------+---------+-----------+--------------------------+------------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
| Projection_4              | 0.80    | 1       | root      |                          | time:213.645µs, loops:2, Concurrency:OFF                               | 1->Column#62                                                                  | 372 Bytes | N/A  |
| └─TableReader_7           | 0.80    | 1       | root      |                          | time:210.966µs, loops:2, rpc num: 1, rpc time:172.033µs, proc keys:0   | data:Selection_6                                                              | 161 Bytes | N/A  |
|   └─Selection_6           | 0.80    | 1       | cop[tikv] |                          | time:48.247µs, loops:2                                                 | concat("0.009009009009009009", cast(bitneg(test.table1_int_autoinc.col_int))) | N/A       | N/A  |
|     └─TableFullScan_5     | 1.00    | 1       | cop[tikv] | table:table1_int_autoinc | time:41.475µs, loops:2                                                 | keep order:false, stats:pseudo                                                | N/A       | N/A  |
+---------------------------+---------+---------+-----------+--------------------------+------------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
4 rows in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------------------+
| Level   | Code | Message                                            |
+---------+------+----------------------------------------------------+
| Warning | 1292 | Truncated incorrect FLOAT value: '18:02:07.032977' |
| Warning | 1292 | Truncated incorrect FLOAT value: '1998-04-26'      |
+---------+------+----------------------------------------------------+
2 rows in set (0.00 sec)
```

WithPushDown:
```
mysql> explain analyze SELECT 1 FROM `table1_int_autoinc` WHERE CONCAT( ( '18:02:07.032977' / '1998-04-26' ), ( ~ `col_int` ) ) ;
+---------------------------+---------+---------+-----------+--------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
| id                        | estRows | actRows | task      | access object            | execution info                                                        | operator info                                                                 | memory    | disk |
+---------------------------+---------+---------+-----------+--------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
| Projection_4              | 0.80    | 0       | root      |                          | time:527.981µs, loops:1, Concurrency:OFF                              | 1->Column#62                                                                  | 372 Bytes | N/A  |
| └─TableReader_7           | 0.80    | 0       | root      |                          | time:527.029µs, loops:1, rpc num: 1, rpc time:472.62µs, proc keys:0   | data:Selection_6                                                              | 241 Bytes | N/A  |
|   └─Selection_6           | 0.80    | 0       | cop[tikv] |                          | time:0s, loops:1                                                      | concat("0.009009009009009009", cast(bitneg(test.table1_int_autoinc.col_int))) | N/A       | N/A  |
|     └─TableFullScan_5     | 1.00    | 1       | cop[tikv] | table:table1_int_autoinc | time:0s, loops:1                                                      | keep order:false, stats:pseudo                                                | N/A       | N/A  |
+---------------------------+---------+---------+-----------+--------------------------+-----------------------------------------------------------------------+-------------------------------------------------------------------------------+-----------+------+
4 rows in set, 3 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------------------------------------------------------+
| Level   | Code | Message                                                                       |
+---------+------+-------------------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect FLOAT value: '18:02:07.032977'                            |
| Warning | 1292 | Truncated incorrect FLOAT value: '1998-04-26'                                 |
| Warning | 1292 | Truncated incorrect INTEGER value: '0.00900900900900900918446744073709551612' |
+---------+------+-------------------------------------------------------------------------------+
3 rows in set (0.00 sec)
```

It seems that the problem is the `Selection_6` executor in Coprocessor.
PTAL @zhongzc

**Comment by SunRunAway** [comment]
A related problem with `UNHEX` function, https://github.com/pingcap/tidb/issues/16631

**Comment by zhongzc** [comment]
Fixed by tikv/tikv/pull/7486
